File: test/unit/serialization/ReferenceObjects.js
/*
 * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

const Long = require('long');
const { UUID } = require('../../../lib/core/UUID');
const { LocalDate, LocalDateTime, LocalTime, OffsetDateTime } = require('../../../lib/core/DateTimeClasses');
const { BigDecimal } = require('../../../lib/core/BigDecimal');
const { HeapData } = require('../../../lib/serialization/HeapData');
const AnInnerPortable = require('./AnInnerPortable');
const AnIdentifiedDataSerializable = require('./AnIdentifiedDataSerializable');
const APortable = require('./APortable');
const CustomByteArraySerializable = require('./CustomSerializable').CustomByteArraySerializable;
const CustomStreamSerializable = require('./CustomSerializable').CustomStreamSerializable;

const to = {};
to.aNULL = null;
to.aBoolean = true;
to.aByte = 113;
to.aCharacter = 'x';
to.aDouble = -897543.3678909;
to.aShort = -500;
to.aFloat = 900.5678;
to.anInteger = 56789;
to.aLong = Long.fromNumber(-50992225);
to.aString = '';
to.aUUID = new UUID(to.aLong, Long.fromNumber(to.anInteger));

for (let ci = 0; ci < 65535; ci++) {
    if (!(ci >= 55296 && ci < 57344)) {
        to.aString += String.fromCharCode(ci);
    }
}
for (let ci = 65535 - to.aString.length; ci > 0; ci--) {
    to.aString += String.fromCharCode(0);
}

to.booleans = [true, false, true];
to.bytes = Buffer.from([112, 4, -1, 4, 112, -35, 43]);
to.chars = ['a', 'b', 'c'];
to.doubles = [-897543.3678909, 11.1, 22.2, 33.3];
to.shorts = [-500, 2, 3];
to.floats = [900.5678, 1.0, 2.1, 3.4];
to.ints = [56789, 2, 3];
to.longs = [Long.fromNumber(-50992225), Long.fromNumber(1231232141), Long.fromNumber(2), Long.fromNumber(3)];
to.Strings = ['Pijamalı hasta, yağız şoföre çabucak güvendi.',
    'イロハニホヘト チリヌルヲ ワカヨタレソ ツネナラム',
    'The quick brown fox jumps over the lazy dog'];
to.AnInnerPortable = new AnInnerPortable(to.anInteger, to.aFloat);
to.aCustomStreamSerializable = new CustomStreamSerializable(to.anInteger, to.aFloat);
to.aCustomByteArraySerializable = new CustomByteArraySerializable(to.anInteger, to.aFloat);

exports.aData = new HeapData(Buffer.from('111313123131313131'));

to.AnIdentifiedDataSerializable = new AnIdentifiedDataSerializable(
    to.aBoolean, to.aByte, to.aCharacter, to.aDouble, to.aShort, to.aFloat,
    to.anInteger, to.aLong, to.aString, to.booleans, to.bytes, to.chars, to.doubles,
    to.shorts, to.floats, to.ints, to.longs, to.Strings, to.AnInnerPortable, null,
    to.aCustomStreamSerializable, to.aCustomByteArraySerializable, exports.aData
);
to.APortable = new APortable(
    to.aBoolean, to.aByte, to.aCharacter, to.aDouble, to.aShort, to.aFloat,
    to.anInteger, to.aLong, to.aString, to.AnInnerPortable, to.booleans, to.bytes,
    to.chars, to.doubles, to.shorts, to.floats, to.ints, to.longs, to.Strings,
    exports.portables, to.AnIdentifiedDataSerializable, to.aCustomStreamSerializable,
    to.aCustomByteArraySerializable, exports.aData
);
to.aDate = new Date(Date.UTC(1990, 2, 1, 0, 0, 0, 0));
to.aBigInteger = BigInt('1314432323232411');
to.aBigDecimal = BigDecimal.fromString('31231');
to.aClass = 'java.math.BigDecimal';

to.aLocalDate = new LocalDate(2021, 6, 28);
to.aLocalTime = new LocalTime(11, 22, 41, 123456789);
to.aLocalDateTime = new LocalDateTime(to.aLocalDate, to.aLocalTime);
to.aOffsetDateTime = new OffsetDateTime(to.aLocalDateTime, 64800);

exports.portables = [to.AnInnerPortable, to.AnInnerPortable, to.AnInnerPortable];
exports.testObjects = to;
exports.skipOnSerialize = {
    'AnIdentifiedDataSerializable': true,
    'APortable': true,
    'aClass': true
};
exports.PORTABLE_FACTORY_ID = 1;
exports.PORTABLE_CLASS_ID = 1;
exports.INNER_PORTABLE_CLASS_ID = 2;
exports.IDENTIFIED_DATA_SERIALIZABLE_FACTORY_ID = 1;
exports.IDENTIFIED_DATA_SERIALIZABLE_CLASS_ID = 1;
exports.CUSTOM_BYTE_ARRAY_SERIALIZABLE_ID = 2;
exports.CUSTOM_STREAM_SERIALIZABLE_ID = 1;

File: test/unit/serialization/ReferenceObjects.java
/*
 * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 This file should be in sync with ReferenceObjects.js file

 When you want to add a new serialization type:

 Update this file and ReferenceObjects.js accordingly. Then use this file to replace
 ReferenceObjects.java in hazelcast repo. Then run BinaryCompatibilityFileGenerator.java to generate the new binary file.
 Then put it here.
*/

package com.hazelcast.nio.serialization.compatibility;

import com.hazelcast.internal.serialization.impl.HeapData;
import com.hazelcast.internal.serialization.Data;
import com.hazelcast.nio.serialization.Portable;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.CharBuffer;
import java.time.LocalTime;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.LinkedList;
import java.util.UUID;

import static java.util.Arrays.asList;

class ReferenceObjects {

    /**
     * PORTABLE IDS
     **/
    static int PORTABLE_FACTORY_ID = 1;
    static int PORTABLE_CLASS_ID = 1;
    static int INNER_PORTABLE_CLASS_ID = 2;

    /**
     * IDENTIFIED DATA SERIALIZABLE IDS
     **/
    static int IDENTIFIED_DATA_SERIALIZABLE_FACTORY_ID = 1;
    static int DATA_SERIALIZABLE_CLASS_ID = 1;

    /**
     * CUSTOM SERIALIZER IDS
     */
    static int CUSTOM_STREAM_SERIALIZABLE_ID = 1;
    static int CUSTOM_BYTE_ARRAY_SERIALIZABLE_ID = 2;

    /**
     * OBJECTS
     */
    static Object aNullObject = null;
    static boolean aBoolean = true;
    static byte aByte = 113;
    static char aChar = 'x';
    static double aDouble = -897543.3678909d;
    static short aShort = -500;
    static float aFloat = 900.5678f;
    static int anInt = 56789;
    static long aLong = -50992225L;
    static String aString;
    static UUID aUUID = new UUID(aLong, anInt);
    static String aSmallString = "😊 Hello Приве́т नमस्ते שָׁלוֹם";

    static {
        CharBuffer cb = CharBuffer.allocate(Character.MAX_VALUE);
        for (char c = 0; c < Character.MAX_VALUE; c++) {
            if (!(c >= Character.MIN_SURROGATE && c < (Character.MAX_SURROGATE + 1))) {
                cb.append(c);
            }
        }
        aString = new String(cb.array());
    }

    static boolean[] booleans = {true, false, true};

    static byte[] bytes = {112, 4, -1, 4, 112, -35, 43};
    static char[] chars = {'a', 'b', 'c'};
    static double[] doubles = {-897543.3678909d, 11.1d, 22.2d, 33.3d};
    static short[] shorts = {-500, 2, 3};
    static float[] floats = {900.5678f, 1.0f, 2.1f, 3.4f};
    static int[] ints = {56789, 2, 3};
    static long[] longs = {-50992225L, 1231232141L, 2L, 3L};
    static String[] strings = {
            "Pijamalı hasta, yağız şoföre çabucak güvendi.",
            "イロハニホヘト チリヌルヲ ワカヨタレソ ツネナラム",
            "The quick brown fox jumps over the lazy dog",
    };

    static Data aData = new HeapData("111313123131313131".getBytes());

    static AnInnerPortable anInnerPortable = new AnInnerPortable(anInt, aFloat);
    static CustomStreamSerializable aCustomStreamSerializable = new CustomStreamSerializable(anInt, aFloat);
    static CustomByteArraySerializable aCustomByteArraySerializable = new CustomByteArraySerializable(anInt, aFloat);
    static Portable[] portables = {anInnerPortable, anInnerPortable, anInnerPortable};

    static AnIdentifiedDataSerializable anIdentifiedDataSerializable = new AnIdentifiedDataSerializable(
            aBoolean, aByte, aChar, aDouble, aShort, aFloat, anInt, aLong, aSmallString,
            booleans, bytes, chars, doubles, shorts, floats, ints, longs, strings,
            anInnerPortable, null,
            aCustomStreamSerializable,
            aCustomByteArraySerializable, aData);
    static APortable aPortable = new APortable(
            aBoolean, aByte, aChar, aDouble, aShort, aFloat, anInt, aLong, aSmallString, anInnerPortable,
            booleans, bytes, chars, doubles, shorts, floats, ints, longs, strings, portables,
            anIdentifiedDataSerializable,
            aCustomStreamSerializable,
            aCustomByteArraySerializable, aData);

    static Date aDate;

    static LocalDate aLocalDate;
    static LocalTime aLocalTime;
    static LocalDateTime aLocalDateTime;
    static OffsetDateTime aOffsetDateTime;

    static {
        Calendar calendar = Calendar.getInstance();
        calendar.set(1990, Calendar.FEBRUARY, 1, 0, 0, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        calendar.set(Calendar.ZONE_OFFSET, 0);
        aDate = calendar.getTime();
        aLocalDate = LocalDate.of(2021, 6, 28);
        aLocalTime = LocalTime.of(11, 22, 41, 123456789);
        aLocalDateTime = LocalDateTime.of(aLocalDate, aLocalTime);
        aOffsetDateTime = OffsetDateTime.of(aLocalDateTime, ZoneOffset.ofHours(18));
    }

    static BigInteger aBigInteger = new BigInteger("1314432323232411");
    static BigDecimal aBigDecimal = new BigDecimal(31231);
    static Class aClass = BigDecimal.class;

    static Object[] objects = {anInnerPortable, aNullObject, aBigDecimal, aShort};

    static ArrayList nonNullList = new ArrayList(asList(
            aBoolean, aByte, aChar, aDouble, aShort, aFloat, anInt, aLong, aSmallString, anInnerPortable,
            booleans, bytes, chars, doubles, shorts, floats, ints, longs, strings,
            aCustomStreamSerializable, aCustomByteArraySerializable,
            anIdentifiedDataSerializable, aPortable,
            aDate, aLocalDate, aLocalTime, aLocalDateTime, aOffsetDateTime, aBigInteger, aBigDecimal, aClass));

    static ArrayList arrayList = new ArrayList(asList(aNullObject, nonNullList));
    static LinkedList linkedList = new LinkedList(arrayList);

    static Object[] allTestObjects = {
            aNullObject, aBoolean, aByte, aChar, aDouble, aShort, aFloat, anInt, aLong, aString, aUUID, anInnerPortable,
            booleans, bytes, chars, doubles, shorts, floats, ints, longs, strings,
            aCustomStreamSerializable, aCustomByteArraySerializable,
            anIdentifiedDataSerializable, aPortable,
            aDate, aLocalDate, aLocalTime, aLocalDateTime, aOffsetDateTime, aBigInteger, aBigDecimal, aClass, objects,
            linkedList, arrayList
    };
}
